File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/message/reactiveobject/ReadChildAttributeResponse.java
package org.opcfoundation.webserver.service.message.reactiveobject;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.UaInstanceNode;
import org.opcfoundation.webserver.addressspace.nodes.UaVariable;

public class ReadChildAttributeResponse {
    private final NodeClass nodeClass;
    private final LocalizedText displayName;
    private final LocalizedText description;
    private final @Nullable NodeId dataTypeId;
    private final @Nullable Integer valueRank;
    private final @Nullable Integer accessLevel;
    private final @Nullable Boolean historizing;

    public ReadChildAttributeResponse(
            NodeClass nodeClass,
            LocalizedText displayName,
            LocalizedText description,
            @Nullable NodeId dataTypeId,
            @Nullable Integer valueRank,
            @Nullable Integer accessLevel,
            @Nullable Boolean historizing)
    {
        this.nodeClass = nodeClass;
        this.displayName = displayName;
        this.description = description;
        this.dataTypeId = dataTypeId;
        this.valueRank = valueRank;
        this.accessLevel = accessLevel;
        this.historizing = historizing;
    }

    public ReadChildAttributeResponse(UaInstanceNode instanceDeclaration)
    {
        this.nodeClass = instanceDeclaration.nodeClass();
        this.displayName = instanceDeclaration.displayName();
        this.description = instanceDeclaration.description();

        if (NodeClass.Variable == instanceDeclaration.nodeClass()) {
            UaVariable variable = (UaVariable)instanceDeclaration;
            this.dataTypeId = variable.dataType();
            this.valueRank = variable.valueRank();
            this.accessLevel = variable.accessLevel();
            this.historizing = variable.historizing();
        } else {
            this.dataTypeId = null;
            this.valueRank = null;
            this.accessLevel = null;
            this.historizing = null;
        }
    }

    public NodeClass getNodeClass() {
        return nodeClass;
    }

    public LocalizedText getDisplayName() {
        return displayName;
    }

    public LocalizedText getDescription() {
        return description;
    }

    public @Nullable NodeId getDataTypeId() {
        return dataTypeId;
    }

    public @Nullable Integer getValueRank() {
        return valueRank;
    }

    public @Nullable Integer getAccessLevel() {
        return accessLevel;
    }

    public @Nullable Boolean getHistorizing() {
        return historizing;
    }
}

File: opcua-webservice-node/src/server/service/message/reactiveobject/ReadChildAttributeResponse.ts
import { NodeClass } from "opcua-webapi";
import { UaLocalizedText, UaNodeId } from "opcua-webapi-ts";
import { UaInstanceNode } from "../../../addressspace/nodes/UaInstanceNode";
import { UaVariable } from "../../../addressspace/nodes/UaVariable";

export class ReadChildAttributeResponse {
    private readonly _nodeClass: NodeClass;
    private readonly _displayName: UaLocalizedText;
    private readonly _description: UaLocalizedText;
    private readonly _dataTypeId: UaNodeId | null;
    private readonly _valueRank: number | null;
    private readonly _accessLevel: number | null;
    private readonly _historizing: boolean | null;

    constructor(
    nodeClass: NodeClass,
    displayName: UaLocalizedText,
    description: UaLocalizedText,
    dataTypeId: UaNodeId | null,
    valueRank: number | null,
    accessLevel: number | null,
        historizing: boolean | null,
    ) {
        this._nodeClass = nodeClass;
    this._displayName = displayName;
    this._description = description;
    this._dataTypeId = dataTypeId;
    this._valueRank = valueRank;
    this._accessLevel = accessLevel;
    this._historizing = historizing;
    }

    get nodeClass(): NodeClass {
        return this._nodeClass;
    }

    get displayName(): UaLocalizedText {
        return this._displayName;
    }

    get description(): UaLocalizedText {
        return this._description;
    }

    get dataTypeId(): UaNodeId | null {
        return this._dataTypeId;
    }

    get valueRank(): number | null {
        return this._valueRank;
    }

    get accessLevel(): number | null {
        return this._accessLevel;
    }

    get historizing(): boolean | null {
        return this._historizing;
    }

    static fromInstanceDeclaration(instanceDeclaration: UaInstanceNode): ReadChildAttributeResponse {
        if (instanceDeclaration.nodeClass === NodeClass.Variable) {
            const variable = instanceDeclaration as UaVariable;

            return new ReadChildAttributeResponse(
                instanceDeclaration.nodeClass,
                instanceDeclaration.displayName,
                instanceDeclaration.description,
                variable.dataType,
                variable.valueRank,
                variable.accessLevel,
                variable.historizing,
            );
        }
        return new ReadChildAttributeResponse(
            instanceDeclaration.nodeClass,
            instanceDeclaration.displayName,
            instanceDeclaration.description,
            null,
            null,
            null,
            null,
        );
    }
}